### PR TITLE
fix(resource-navigation): query exact latest and reusable resources

### PR DESCRIPTION
**Pull request by zhangjiadi225**
> ### Branch strategy
>
> - Active development targets `main`.
> - v1 maintenance targets `v1`; forward-port fixes to `main` separately when needed.

### What this PR does

Before this PR:

- Classic-layout assistant and agent rails selected the latest resource from the currently loaded client list, so pagination or stale cache state could open the wrong topic or session.
- Empty placeholder reuse scanned client state and, for agent sessions, issued up to eight message probes before deciding whether a session was reusable.
- An older asynchronous owner lookup or post-deletion fallback could overwrite a newer selection or route change.

After this PR:

- The latest-topic and latest-session endpoints accept an exact owner scope, including unlinked resources, and query activity ordering directly in SQLite.
- New transactional endpoints atomically reuse or create structurally empty placeholders for an exact assistant/agent and workspace target.
- Agent placeholder reuse uses a database anti-join for message emptiness and cleans duplicate empty system sessions in the same transaction.
- Classic-layout owner activation and post-deletion fallback ignore stale asynchronous results, preserving the user's newest selection.
- Client-side candidate scans, bounded message probes, and duplicate-session deletion are removed from the page components.

Fixes # N/A

### Why we need it and why it was done in this way

The loaded resource list is a presentation cache, not an authoritative latest-resource index. Moving exact selection and placeholder eligibility into the SQLite-backed data services makes the behavior correct across pagination, stale renderer state, repeated requests, and concurrent UI actions. It also replaces multiple client-side message probes with one API request and one serialized database transaction.

The following tradeoffs were made:

- Selecting an assistant or agent now performs one exact backend lookup instead of immediately trusting the in-memory list. This adds a request boundary but avoids incorrect navigation and scales independently of the loaded page.
- The existing full-list source remains in place for rendering the rail, so this PR does not change initial list-loading cost.
- Placeholder reuse stays explicit in the DataApi rather than being folded into ordinary create endpoints, preserving the semantics of callers that require a genuinely fresh resource.

The following alternatives were considered:

- Continue sorting the loaded renderer list and probing candidate sessions for messages — rejected because the list can be incomplete and the probes add bounded but unnecessary I/O.
- Add more client-side cache invalidation and duplicate cleanup — rejected because it cannot make reuse atomic across repeated requests.

Links to places where the discussion took place: https://github.com/CherryHQ/cherry-studio/pull/16434

### Breaking changes

If this PR introduces breaking changes, please describe the changes and the impact on users.

None.

### Special notes for your reviewer

- `pnpm lint` passed, including typecheck, i18n validation, and formatting. ESLint reported 40 pre-existing warnings outside this change set and no errors.
- Commit hooks passed Biome and ESLint checks.
- Tests, `pnpm build:check`, and manual UI verification were not run, per the workspace verification policy.
- The branch is rebased onto the latest `origin/main`; the single commit is cryptographically signed and DCO-signed off.

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [x] Branch: This PR targets the correct branch — `main` for active development, `v1` for v1 maintenance fixes
- [x] PR: The PR description is expressive enough and will help future contributors
- [x] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [x] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [x] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [x] Documentation: A [user-guide update](https://docs.cherry-ai.com) was considered and is present (link) or not required. Check this only when the PR introduces or changes a user-facing feature or behavior.
- [x] Self-review: I have reviewed my own code (e.g., via [`/gh-pr-review`](/.claude/skills/gh-pr-review/SKILL.md), `gh pr diff`, or GitHub UI) before requesting review from others

### Release note

```release-note
Classic-layout assistant and agent navigation now opens the exact latest conversation or work session, reuses empty placeholders without duplicate creation, and no longer lets stale asynchronous lookups override a newer selection.
```

### CODE REVIEW

File: src/renderer/components/chat/resourceList/useOwnerResourceActivation.ts [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+}: UseOwnerResourceActivationOptions<TOwner, TResource>) {
+  const requestGenerationRef = useRef(0)
+
+  const cancelOwnerResourceActivation = useCallback(() => {
+    requestGenerationRef.current += 1
+  }, [])
+
+  useEffect(
+    () => cancelOwnerResourceActivation,
+    [cancelOwnerResourceActivation, createResourceForOwner, loadResourceForOwner]
```
[the remark this answers is not in the document]
Comment (zhangjiadi225) [reply src/renderer/components/chat/resourceList/useOwnerResourceActivation.ts]: Fixed in `028b251291`. The cleanup effect now depends only on the stable cancel callback, so a parent rerender that changes load/create callback identity no longer increments the request generation or drops a valid in-flight selection. I also added a regression test that starts a deferred owner lookup, rerenders with a new create callback, and verifies that the selected resource is still activated. `pnpm lint` and `pnpm format` pass; targeted tests were not run under this workspace verification policy.

[the remark this answers is not in the document]
Comment (zhangjiadi225) [reply src/renderer/components/chat/resourceList/useOwnerResourceActivation.ts]: Addressed by the same fix in `028b251291`. The agent-side callback churn described here no longer triggers the unmount cleanup. Actual entity changes still cancel through the explicit `activeEntityId` effect, and newer clicks still supersede older requests through the generation guard. The regression test covers callback churn during an in-flight lookup.

**Comment by zhangjiadi225** [comment]
Thanks for flagging the pin-sync concern. I rechecked the commit graph: `09d8159` and the original PR commit `ebac8870` are sibling commits with common parent `94a0cf9`; the PR-side diff from that merge base to `ebac8870` does not modify `PinService.ts` or `usePins.ts`. The apparent deletions come from comparing the two complete snapshots (`09d8159..ebac8870`), which includes base-only changes as deletions. The PR branch has since merged current `main`, and the current head retains `notifyPinReadModelChange`, `notifyPurged`, the purge call sites, and `useDataChange('/pins')`. Therefore this PR does not remove cross-window pin synchronization, and no pin-specific code change is needed.

**Comment by zhangjiadi225** [comment]
> This comment was translated automatically.

@claude please re-review

---
<details>
<summary>Original Content</summary>

@claude 复审一下
</details>

File: src/main/data/services/AgentSessionService.ts [hunk 2]
```diff
@@ -357,18 +360,126 @@ export class AgentSessionService {
...
+            taskScheduleIds: []
+          }
+        }),
+      {
+        ...defaultHandlersFor('Session', reservedId),
+        foreignKey: () => DataApiErrorFactory.notFound('Agent or Workspace')
+      }
+    )
+
+    publishTaskReadModelChanges(result.taskScheduleIds)
```
Comment (AtomsH4) [inline src/main/data/services/AgentSessionService.ts:471]: > This review comment was translated automatically.

**[Warning · A1]** The system-placeholder deduplication here goes through `cascadeDeleteSessionRowsTx()`, which deletes the duplicate session's pins via `pinService.purgeForEntitiesTx()`; however, after this new wrapper commits, it only publishes the task/session read models, without calling `pinService.notifyPurged()` like existing deletion entry points such as `delete()` / `deleteByIds()`. The local invalidation in `AgentPage.resolveEmptySession()` also does not include `/pins`, so if the purged empty session was previously pinned, the current window and other windows will continue to retain stale `/pins` cache until an unrelated next revalidation. Please call `pinService.notifyPurged()` after the commit when `deletedDuplicateSessionIds.length > 0`, and add a regression test that first pins a duplicate session, then asserts the `/pins` membership notification.

---
<details>
<summary>Original Content</summary>

**[Warning · A1]** 这里的 system-placeholder 去重会走 `cascadeDeleteSessionRowsTx()`，后者会通过 `pinService.purgeForEntitiesTx()` 删除重复 session 的 pin；但这个新 wrapper 提交后只发布 task/session 读模型，没有像 `delete()` / `deleteByIds()` 等既有删除入口那样调用 `pinService.notifyPurged()`。`AgentPage.resolveEmptySession()` 的本地 invalidation 也不包含 `/pins`，因此如果被清理的空 session 曾被 pin，当前窗口和其他窗口会继续保留过期的 `/pins` 缓存，直到无关的下一次 revalidation。请在提交后按 `deletedDuplicateSessionIds.length > 0` 调用 `pinService.notifyPurged()`，并补一个会先 pin duplicate session、再断言 `/pins` membership 通知的回归测试。
</details>

**Review by AtomsH4 — COMMENTED** [review]
> This review was translated automatically.

Overall change direction aligns with `CLAUDE.md`: latest resource and placeholder reuse is determined by the SQLite-owned service, and read/write transaction and renderer race protection also follow repository practices. No style or component API changes were made, and no `DESIGN.md` violations were found. 1 Warning found: agent placeholder deduplication misses the pins read model notification, which may cause the current window and other windows to retain stale pins. Please fix before considering approval. Non-linear commit history is not a blocking item.

---
<details>
<summary>Original Content</summary>

总体改动方向符合 `CLAUDE.md`：最新资源与占位复用由 SQLite 所属服务判定，读写事务和 renderer 的竞态保护也符合仓库实践；本次没有样式或组件 API 改动，未发现 `DESIGN.md` 违例。发现 1 个 Warning：agent placeholder 去重遗漏 pins 读模型通知，可能让当前窗口及其他窗口保留过期 pin，请修复后再考虑 approve。非线性提交记录未作为阻塞项。
</details>

File: src/main/data/services/AgentSessionService.ts (hunk 2 above)
Comment (zhangjiadi225) [reply src/main/data/services/AgentSessionService.ts:471]: Fixed in 49521af368. The reuse-or-create wrapper now calls pinService.notifyPurged() after the transaction whenever duplicate system placeholders were deleted. I added a database regression test that pins the duplicate, verifies the pin row is purged, and asserts the /pins membership notification. I also updated useSessions to refetch its session-pin query on /pins notifications so the originating window and other windows converge. pnpm lint and pnpm format pass; targeted tests and build:check were not run under this workspace verification policy.

**Review by AtomsH4 — APPROVED** [review]
> This review was translated automatically.

Review approved. The previously raised `/pins` read model sync issue has been fixed in `49521af368`: after duplicate system placeholder removal, a pin membership notification is published, with database and renderer regression tests covering pin cleanup/notification and cross-window refetch respectively. No new Blockers or Warnings found; the change direction still aligns with `CLAUDE.md`, the repository's DataApi practices, and `DESIGN.md`. All CI passed; non-linear commit history was not treated as a blocking item.

---
<details>
<summary>Original Content</summary>

复审通过。上次提出的 `/pins` 读模型同步问题已在 `49521af368` 修复：重复 system placeholder 删除后会发布 pin membership 通知，数据库与 renderer 回归测试分别覆盖 pin 清理/通知和跨窗口重取。未发现新的 Blocker 或 Warning；改动方向仍符合 `CLAUDE.md`、仓库 DataApi 实践及 `DESIGN.md`。CI 全部通过，非线性提交记录未作为阻塞项。
</details>

**Review by eeee0717 — APPROVED** [review]
LGTM
